pennylane/numpy/tensor.py

```diff
@@ -132,6 +132,10 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
         # unwrap the input arguments to the ufunc
         args = [i.unwrap() if hasattr(i, "unwrap") else i for i in inputs]
 
+        for i in args:
+            if hasattr(i, "unbox"):
+                return i.__array_ufunc__(ufunc, method, *args, **kwargs)

```

Comment:
> Since the return statement will bypass the remainder of the for loop

This was somewhat deliberate. Here, I'm checking if any of the arrays are `TensorBox` objects implicitly (I cannot import the tensorbox module here as the NumPy module is much lower down the hierarchy, and it would cause circular imports). If an object in the input _is_ a tensor box, that we instead send this ufunc request to The `TensorBox.__array_ufunc__` method.

This is an issue with how NumPy dispatches a `ufunc`. If multiple arguments of the ufunc define custom `__array_ufunc__` methods, then we cannot know for sure which one will be called (it depends on the order of the input). See the discussion in https://numpy.org/devdocs/user/basics.subclassing.html#array-ufunc-for-ufuncs:

> E.g., lets assume that we evalulate `np.add(a, b)`, where b is an instance of another class B that has an override. If you use super as in the example, `ndarray.__array_ufunc__` will notice that b has an override, which means it cannot evaluate the result itself. Thus, it will return `NotImplemented` and so will our class A. Then, control will be passed over to b, which either knows how to deal with us and produces a result, or does not and returns `NotImplemented`, raising a `TypeError`.

Reply:
I think I've got it, thanks @josh146